nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java

```diff
@@ -0,0 +1,499 @@
+        flowFileMaxFailure = context.getProperty(FF_FAILURE_ITERATIONS).asInteger();
+        flowFileMaxYield = context.getProperty(FF_ROLLBACK_YIELD_ITERATIONS).asInteger();
+        flowFileMaxRollback = context.getProperty(FF_ROLLBACK_ITERATIONS).asInteger();
+        flowFileMaxPenalty = context.getProperty(FF_ROLLBACK_PENALTY_ITERATIONS).asInteger();
+        flowFileMaxException = context.getProperty(FF_EXCEPTION_ITERATIONS).asInteger();
+        noFlowFileMaxException = context.getProperty(NO_FF_EXCEPTION_ITERATIONS).asInteger();
+        noFlowFileMaxYield = context.getProperty(NO_FF_YIELD_ITERATIONS).asInteger();
+        noFlowFileMaxSkip = context.getProperty(NO_FF_SKIP_ITERATIONS).asInteger();
+        curr_ff_resp.reset();
+        curr_noff_resp.reset();
+        flowFileExceptionClass = classNameToRuntimeExceptionClass(context.getProperty(FF_EXCEPTION_CLASS).toString());
+        noFlowFileExceptionClass = classNameToRuntimeExceptionClass(context.getProperty(NO_FF_EXCEPTION_CLASS).toString());
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        final ComponentLog logger = getLogger();
+
+        FlowFile ff = session.get();
+
+        // Make up to 2 passes to allow rollover from last cycle to first.
+        // (This could be "while(true)" since responses should break out  if selected, but this
+        //  prevents endless loops in the event of unexpected errors or future changes.)
+        int pass = 2;
+        while (pass > 0) {
+            pass -= 1;
+            if (ff == null) {
+                if (curr_noff_resp.state() == NoFlowFileResponseState.NO_FF_SKIP_RESPONSE) {
+                    if (noFlowFileCurrSkip < noFlowFileMaxSkip) {
+                        noFlowFileCurrSkip += 1;
+                        logger.info("DebugFlow skipping with no flow file");
+                        return;
+                    } else {
+                        noFlowFileCurrSkip = 0;
+                        curr_noff_resp.getNextCycle();
+                    }
+                }
+                if (curr_noff_resp.state() == NoFlowFileResponseState.NO_FF_EXCEPTION_RESPONSE) {
+                    if (noFlowFileCurrException < noFlowFileMaxException) {
+                        noFlowFileCurrException += 1;
+                        logger.info("DebugFlow throwing NPE with no flow file");
+                        String message = "forced by " + this.getClass().getName();
+                        RuntimeException rte;
+                        try {
+                            rte = noFlowFileExceptionClass.getConstructor(String.class).newInstance(message);
+                            throw rte;
+                        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+                            e.printStackTrace();
```

Comment:
Validation will have verified that the class exists and can be loaded by the processor, so this should (haha) never occur.  I moved it to a debug level log message.